internal/controllers/machineset/machineset_controller.go

```diff
@@ -359,6 +329,79 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 	return result, nil
 }
 
+func (r *Reconciler) reconcileDelete(ctx context.Context, machineSet *clusterv1.MachineSet) error {
+	log := ctrl.LoggerFrom(ctx)
+	machineList, err := r.getAndAdoptMachinesForMachineSet(ctx, machineSet)
+	if err != nil {
+		return err
+	}
+
+	// If all the descendant machines are deleted, then remove the machineset's finalizer.
+	if len(machineList) == 0 {
+		controllerutil.RemoveFinalizer(machineSet, clusterv1.MachineSetFinalizer)
+		return nil
+	}
+
+	log.Info("Waiting for Machines to be deleted", "Machines", clog.ObjNamesString(machineList))
+
+	// else delete owned machines.
+	for _, machine := range machineList {
+		if machine.DeletionTimestamp.IsZero() {
+			log.Info("Deleting Machine", "Machine", klog.KObj(machine))
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return errors.Wrapf(err, "failed to delete Machine %s", klog.KObj(machine))
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *Reconciler) getAndAdoptMachinesForMachineSet(ctx context.Context, machineSet *clusterv1.MachineSet) ([]*clusterv1.Machine, error) {
+	log := ctrl.LoggerFrom(ctx)
+	selectorMap, err := metav1.LabelSelectorAsMap(&machineSet.Spec.Selector)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to convert MachineSet %q label selector to a map", machineSet.Name)
+	}
+
+	// Get all Machines linked to this MachineSet.
+	allMachines := &clusterv1.MachineList{}
+	err = r.Client.List(ctx,
+		allMachines,
+		client.InNamespace(machineSet.Namespace),
+		client.MatchingLabels(selectorMap),
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list machines")
+	}
+
+	// Filter out irrelevant machines (i.e. IsControlledBy something else) and claim orphaned machines.
+	// Machines in deleted state are deliberately not excluded https://github.com/kubernetes-sigs/cluster-api/pull/3434.
+	filteredMachines := make([]*clusterv1.Machine, 0, len(allMachines.Items))
+	for idx := range allMachines.Items {
+		machine := &allMachines.Items[idx]
+		log := log.WithValues("Machine", klog.KObj(machine))
```

Comment:
```suggestion
		log := log.WithValues("Machine", klog.KObj(machine))
		ctx := ctrl.LoggerInto(ctx, log)
```
I know it was the same before, but let's improve